File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    public long solution(int[] sequence) {
        long answer = 0;
        
        long[] dp1 = new long[sequence.length];
        long[] dp2 = new long[sequence.length];
        
        int a = 1;
        int b = -1;
        
        dp1[0] = sequence[0] * a;
        dp2[0] = sequence[0] * b;
        
        answer = Math.max(dp1[0], dp2[0]);
        
        for(int i = 1; i < sequence.length; i++){
            a *= -1;
            b *= -1;
            
            int next1 = sequence[i] * a;
            int next2 = sequence[i] * b;
            
            
            dp1[i] = Math.max(dp1[i-1] + next1, next1);
            dp2[i] = Math.max(dp2[i-1] + next2, next2);
            
            answer = Math.max(answer, Math.max(dp1[i], dp2[i]));
        }
        
        return answer;
    }
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
    let answer = 0;
    
    let dpPM = new Array(sequence.length).fill(-Infinity);
    let dpMP = new Array(sequence.length).fill(-Infinity);
    
    let PM = +1;
    let MP = -1;
    
    dpPM[0] = sequence[0] * PM;
    dpMP[0] = sequence[0] * MP;
    
    for(let i = 0; i < sequence.length; i++){
        if(i > 0) {
            dpPM[i] = Math.max(dpPM[i-1] + sequence[i] * PM, sequence[i] * PM);
            dpMP[i] = Math.max(dpMP[i-1] + sequence[i] * MP, sequence[i] * MP);
        }        
        
        PM *= -1;
        MP *= -1;
        
        answer = Math.max(answer, Math.max(dpPM[i], dpMP[i]));
    }
    
    return answer;
}